src/toxcore/tcp/packet.rs

```diff
@@ -457,6 +466,168 @@ impl ToBytes for OobReceive {
     }
 }
 
+
+
+
+
+
+
+// TODO: move it somewhere else
+impl ToBytes for IpAddr {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        match *self {
+            IpAddr::V4(ref p) => p.to_bytes(buf),
+            IpAddr::V6(ref p) => p.to_bytes(buf),
+        }
+    }
+}
+
+// TODO: move it somewhere else
+impl FromBytes for Ipv4Addr {
+    named!(from_bytes<Ipv4Addr>, map!(count!(le_u8, 4), 
+        |v| Ipv4Addr::new(v[0], v[1], v[2], v[3])
+    ));
+}
+
+impl ToBytes for Ipv4Addr {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        let o = self.octets();
+        do_gen!(buf,
+            gen_be_u8!(o[0]) >>
+            gen_be_u8!(o[1]) >>
+            gen_be_u8!(o[2]) >>
+            gen_be_u8!(o[3]) 
+        )
+    }
+}
+
+// TODO: move it somewhere else
+impl FromBytes for Ipv6Addr {
+    named!(from_bytes<Ipv6Addr>, map!(count!(le_u16, 8), 
+        |v| Ipv6Addr::new(v[0], v[1], v[2], v[3], v[4], v[5], v[6], v[7])
+    ));
+}
+
+impl ToBytes for Ipv6Addr {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        let s = self.segments();
+        do_gen!(buf,
+            gen_le_u16!(s[0]) >>
+            gen_le_u16!(s[1]) >>
+            gen_le_u16!(s[2]) >>
+            gen_le_u16!(s[3]) >>
+            gen_le_u16!(s[4]) >>
+            gen_le_u16!(s[5]) >>
+            gen_le_u16!(s[6]) >>
+            gen_le_u16!(s[7]) 
+        )
+    }
+}
+
+
+
+
+
+/// IPv4 is padded with 12 bytes of zeroes so that both IPv4 and
+/// IPv6 have the same stored size.
+pub const IPV4_PADDING_SIZE: usize = 12;
+
+// TODO: move to onion?
+#[derive(Debug, PartialEq, Clone)]
+pub struct IpPort {
+    pub addr: IpAddr,
+    pub port: u16
+}
+
+impl FromBytes for IpPort {
+    named!(from_bytes<IpPort>, do_parse!(
+        ip_type: le_u8 >>
+        addr: alt!(
+            cond_reduce!(ip_type == 2 || ip_type == 130, terminated!(
```

Comment:
In TCP Relay we must strictly accept only TCP4 and TCP6

Reply:
Not exactly. They have bool flag disable_family_check (see onion.c#65) and set it to true when parse IP for TCP server. So TCP server can accept both TCP and UDP families.